source/azure_iot_hub_client.c

```diff
@@ -626,3 +740,527 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    uint16_t usSubscribePacketIdentifier;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to unsubscribe device twin: Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_TWIN ];
+        xMqttSubscription[ 0 ].xQoS = eAzureIoTMQTTQoS0;
+        xMqttSubscription[ 0 ].pcTopicFilter = ( const uint8_t * ) AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_SUBSCRIBE_TOPIC;
+        xMqttSubscription[ 0 ].usTopicFilterLength = ( uint16_t ) sizeof( AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_SUBSCRIBE_TOPIC ) - 1;
+        xMqttSubscription[ 1 ].xQoS = eAzureIoTMQTTQoS0;
+        xMqttSubscription[ 1 ].pcTopicFilter = ( const uint8_t * ) AZ_IOT_HUB_CLIENT_TWIN_PATCH_SUBSCRIBE_TOPIC;
+        xMqttSubscription[ 1 ].usTopicFilterLength = ( uint16_t ) sizeof( AZ_IOT_HUB_CLIENT_TWIN_PATCH_SUBSCRIBE_TOPIC ) - 1;
+        usSubscribePacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+
+        AZLogDebug( ( "Attemptting to unsubscribe from the MQTT  %s and %s topics.",
+                      AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_SUBSCRIBE_TOPIC, AZ_IOT_HUB_CLIENT_TWIN_PATCH_SUBSCRIBE_TOPIC ) );
+
+        if( ( xMQTTResult = AzureIoTMQTT_Unsubscribe( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                      xMqttSubscription, 2,
+                                                      usSubscribePacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to unsubscribe, error=0x%08x", xMQTTResult ) );
+            xReturn = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            memset( pxContext, 0, sizeof( AzureIoTHubClientReceiveContext_t ) );
+            xReturn = eAzureIoTHubClientSuccess;
+        }
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SendDeviceTwinReported( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                                    const uint8_t * pucReportedPayload,
+                                                                    uint32_t ulReportedPayloadLength,
+                                                                    uint32_t * pulRequestId )
+{
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    AzureIoTMQTTPublishInfo_t xMQTTPublishInfo = { 0 };
+    uint8_t ucRequestID[ azureiothubMAX_SIZE_FOR_UINT32 ];
+    size_t xTopicLength;
+    az_result xCoreResult;
+    az_span xRequestID = az_span_create( ucRequestID, sizeof( ucRequestID ) );
+
+    if( ( pxAzureIoTHubClient == NULL ) ||
```

Comment:
should we be checking `pulRequestID` here?

Reply:
it is optional

---
In reply to: [612835601](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612835601) [](ancestors = 612835601)